framework/Optimizers/SimulatedAnnealing.py

```diff
@@ -0,0 +1,718 @@
+      self._CoolingParameters['beta'] = 0.1
+      self._CoolingParameters['c'] = 1.0
+      self._CoolingParameters['d'] = 1.0
+
+    type = self._coolingMethod
+    if type == 'linear':
+      beta = self._CoolingParameters['beta']
+      return T0 - iter * beta
+    elif type in ['exponential','geometric']:
+      alpha = self._CoolingParameters['alpha']
+      return alpha ** iter * T0
+    elif type == 'boltzmann':
+      d = self._CoolingParameters['d']
+      return T0/(np.log10(iter + d))
+    elif type == 'fast':
+      c = self._CoolingParameters['c']
+      return np.exp(-c*iter) * T0
+    elif type == 'veryfast':
+      c = self._CoolingParameters['c']
+      return np.exp(-c*iter**(1/len(self.toBeSampled.keys()))) * T0
+    elif type == 'cauchy':
+      d = self._CoolingParameters['d']
+      return T0/(iter + d)
+    else:
+      raise NotImplementedError('cooling schedule type not implemented.')
+
+  def _nextNeighbour(self, rlz,fraction=1,alpha = 0.94,cont = True):
+    """ Perturb x to find the next random neighbour
+        for linear and exponential cooling:
+        .. math::
+
+            fraction = \\frac{iter}{Limit}
+
+            amp = 1-fraction
+
+            delta = \\frac{-amp}{2} + amp * r
+        where :math: `r \sim \mathcal{U}(0,1)`
+
+        for boltzmann cooling:
+        .. math::
+
+            amp = min(\\sqrt(T), \\frac{1}{3*alpha}
+
+            delta = r * alpha * amp
+
+        where :math: `r \\sim \\mathcal{N}(0,1)`
+
+        for fast cooling:
+        .. math::
+
+            amp = r
+
+            delta = sign(amp-0.5)*T*((1+\\frac{1.0}{T})^{\\abs{2*amp-1}-1.0)
+
+        where :math: `r \\sim \\mathcal{U}(0,1)`
+
+        for cauchy cooling:
+        .. math::
+
+            amp = r
+
+            delta = alpha * T * tan(amp)
+
+        where :math: `r \\sim \\mathcal{U}(-\\pi,\\pi)`
+    """
+    if cont:
+      nextNeighbour = {}
+      D = len(self.toBeSampled.keys())
+      if self._coolingMethod in ['linear' , 'exponential', 'geometric']:
+        amp = ((fraction)**-1) / 20
+        r = randomUtils.random(dim=D, samples=1)
+        delta = (-amp/2.)+ amp * r
+      elif self._coolingMethod == 'boltzmann':
+        amp = min(np.sqrt(self.T), 1/3.0/alpha)
+        delta =  randomUtils.randomNormal(dim=D, samples=1)*alpha*amp
+      elif self._coolingMethod in ['fast','veryfast']:
+        amp = randomUtils.random(dim=D, samples=1)
+        T = self.T
+        delta = np.sign(amp-0.5)*T*((1+1.0/T)**abs(2*amp-1)-1.0)
+      elif self._coolingMethod == 'cauchy':
+        amp = (np.pi - (-np.pi))*randomUtils.random(dim=D, samples=1)-np.pi
+        delta = alpha*self.T*np.tan(amp)
+      for i,var in enumerate(self.toBeSampled.keys()):
+        nextNeighbour[var] = rlz[var] + delta[i]
+        self.info['amp_'+var] = amp
+        self.info['delta_'+var] = delta[i]
+      self.info['fraction'] = fraction
+    else:
+      # This is for list-baed SA
+      # TODO
+      pass
```

Comment:
?

Reply:
you can put a TODO comment...but not an if statement without (basically) an ```else```

Reply:
Done, converted to a TODO